### PR TITLE
Expose indices with non-active primary shards in publications

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Fixes https://github.com/crate/crate/issues/17734

Reverts https://github.com/crate/crate/commit/b6a8c5ed02730058b9eb53f7de0390b679a64dbc as it introduced a regression.

If we don't expose some indices, a subscriber cluster thinks that index is gone and stops tracking. See https://github.com/crate/crate/commit/40c12fb23d2dbace52a951e714ccc0189c795281 for details.

This change goes with an alternative way - publisher keeps exposing all relevant indices and subscriber keeps polling shard changes. However, publisher throws an error so that subscriber can keep the index and retry later.

Redundant exposure lasts only for a short time window (typically after the publisher restart) when primaries are not yet assigned.

### CODE REVIEW

File: server/src/main/java/io/crate/replication/logical/action/ShardChangesAction.java [hunk 1]
```diff
@@ -231,6 +237,34 @@ protected ShardsIterator shards(ClusterState state, Request request) {
...
+         * For example, this can happen on a publisher cluster restart.
+         *
+         * @return boolean flag indicating that listener is failed and async callee can return right away.
+         */
+        private boolean ensureAllPrimariesActive(ClusterState state,
+                                                 ShardId shardId,
+                                                 @Nullable ActionListener<Response> listener) {
+            var routingTable = state.routingTable().index(shardId.getIndexName());
+            assert routingTable != null : "routingTable must not be null";
+            if (routingTable.allPrimaryShardsActive() == false) {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/replication/logical/action/ShardChangesAction.java:256]: It's WIP because some tests are failing with an assertion error. If I add NPE check, test pass but I want to fugure out why we can have null routing table. 

Also, cluster state is completely empty when `routingTable` is null, so maybe test setup issue

**Comment by BaurzhanSakhariev** [comment]
Another alternative is to change publication response to expose everything but differentiate between "dropped" and "not yet ready to publish".

It can help to avoid some extra polling but would need to change publisher/subscriber streaming (not a big deal) but also need to change `MetadataTracker` to react only to actually dropped tables (based on some extra flag) and propagate info about "not yet ready" indices to the `ShardReplicationChangesTracker` which I found error prone and not worthwhile.

Time window when not all shards are active yet should be short.

**Comment by BaurzhanSakhariev** [comment]
test_deleted_and_recreated_partition_is_also_deleted_and_restored_on_subscriber is still failing, even with NPE check on `routingTable`

**Comment by BaurzhanSakhariev** [comment]
Due to a table being hold on on a publisher and subscriber thinking it got removed, 

we run into the check introduced in https://github.com/crate/crate/pull/12342 (not the root cause, side effect).

I added some logging to the `MetadataTracker.getExistingLocallyTables` and `MetadataTracker.processDroppedTablesOrPartitions`

On master
```
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] currentlyReplicatedTables : [doc.x]
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] publisher response indices: [] -> publisher has unassigned primaries after restarts, doesn't send doc.x for a while
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] existing relations (must be empty): [] - all good, `getExistingLocallyTables` check passes this iteration
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] changedRelations (will be tuned into regular writable/droppable tables): [doc.x] --> `processDroppedTablesOrPartitions` kicks in
[INFO ][i.c.r.l.ShardReplicationChangesTracker] [Tschingelhorn] [[x][0]] Temporary error during renewal of retention leases for subscription 'rs'. Retrying: NoShardAvailableActionException:No shard available for [org.elasticsearch.index.seqno.RetentionLeaseActions$AddOrRenewRequest@6d29b0f8]
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] currentlyReplicatedTables : [] --> doc.x is not being replicated
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] publisher response indices: [] -> still no active primaries
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] existing relations (must be empty): []
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] currentlyReplicatedTables : []
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] publisher response indices: []
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] existing relations (must be empty): []
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] currentlyReplicatedTables : [] --> doc.x is not being replicated
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] publisher response indices: [x] -> publisher finally got active primaries after restart
[INFO ][i.c.r.l.MetadataTracker  ] [Tschingelhorn] existing relations (must be empty): [doc.x] -- check fails
[ERROR][i.c.r.l.MetadataTracker  ] [Tschingelhorn] Tracking of metadata failed for subscription 'rs', stopping tracking. Some relation(s) already exist. Check table pg_subscription_rel to see existing tables.
```

With this PR (currentlyReplicatedTables : constantly has [doc.x], we have couple of rounds where tracker sees `NoShardAvailableActionException` but eventually (when primaries are active) we are back to the state when there are no exception in logs


```
Temporary error during renewal of retention leases for subscription 'rs'. Retrying: ConnectTransportException:[][127.0.0.1:4300] general node connection failure
[2025-04-25T16:34:45,289][INFO ][i.c.r.l.ShardReplicationChangesTracker] [Monte Cernera] [[x][0]] Temporary error during renewal of retention leases for subscription 'rs'. Retrying: ClusterBlockException:blocked by: [SERVICE_UNAVAILABLE/1/state not recovered / initialized];
[2025-04-25T16:34:45,349][INFO ][i.c.r.l.ShardReplicationChangesTracker] [Monte Cernera] [[x][0]] Temporary error during renewal of retention leases for subscription 'rs'. Retrying: ClusterBlockException:blocked by: [SERVICE_UNAVAILABLE/1/state not recovered / initialized];
[2025-04-25T16:34:45,376][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] currentlyReplicatedTables : [doc.x]
[2025-04-25T16:34:45,376][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] publisher response indices: [x]
[2025-04-25T16:34:45,376][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] existing relations (must be empty): []
[2025-04-25T16:34:45,408][INFO ][i.c.r.l.ShardReplicationChangesTracker] [Monte Cernera] [[x][0]] Temporary error during renewal of retention leases for subscription 'rs'. Retrying: NoShardAvailableActionException:No shard available for [org.elasticsearch.index.seqno.RetentionLeaseActions$AddOrRenewRequest@106c7772]
[2025-04-25T16:34:45,435][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] currentlyReplicatedTables : [doc.x]
[2025-04-25T16:34:45,436][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] publisher response indices: [x]
[2025-04-25T16:34:45,436][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] existing relations (must be empty): []
[2025-04-25T16:34:45,466][INFO ][i.c.r.l.ShardReplicationChangesTracker] [Monte Cernera] [[x][0]] Temporary error during renewal of retention leases for subscription 'rs'. Retrying: NoShardAvailableActionException:No shard available for [org.elasticsearch.index.seqno.RetentionLeaseActions$AddOrRenewRequest@85cf557]


[2025-04-25T16:34:45,494][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] currentlyReplicatedTables : [doc.x]
[2025-04-25T16:34:45,494][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] publisher response indices: [x]
[2025-04-25T16:34:45,495][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] existing relations (must be empty): []
[2025-04-25T16:34:45,526][INFO ][i.c.r.l.ShardReplicationChangesTracker] [Monte Cernera] [[x][0]] Temporary error during renewal of retention leases for subscription 'rs'. Retrying: NoShardAvailableActionException:No shard available for [org.elasticsearch.index.seqno.RetentionLeaseActions$AddOrRenewRequest@5c85779b]
==> primary is again active after restart, we are back to the good state, subscriber didn't lose doc.x and publisher keeps sending it.
[2025-04-25T16:34:45,557][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] currentlyReplicatedTables : [doc.x]
[2025-04-25T16:34:45,557][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] publisher response indices: [x]
[2025-04-25T16:34:45,561][INFO ][i.c.r.l.MetadataTracker  ] [Monte Cernera] existing relations (must be empty): []
```

**Comment by BaurzhanSakhariev** [comment]
> `test_deleted_and_recreated_partition_is_also_deleted_and_restored_on_subscriber` is still failing, even with NPE check on `routingTable`

Well, because, initial restore is failing. 

Looks like have to go with initial approach (https://github.com/crate/crate/pull/17818#issuecomment-2835231905)

File: server/src/main/java/io/crate/replication/logical/action/ShardChangesAction.java [hunk 2]
```diff
@@ -231,6 +237,33 @@ protected ShardsIterator shards(ClusterState state, Request request) {
...
+         * can treat "not ready to replicate" index as deleted and turn a table into a regular one on its side.
+         * For example, this can happen on a publisher cluster restart.
+         *
+         * @return boolean flag indicating that listener is failed and async callee can return right away.
+         */
+        private boolean ensureAllPrimariesActive(ClusterState state,
+                                                 ShardId shardId,
+                                                 @Nullable ActionListener<Response> listener) {
+            var routingTable = state.routingTable().index(shardId.getIndexName());
+            if (routingTable != null && routingTable.allPrimaryShardsActive() == false) {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/replication/logical/action/ShardChangesAction.java:255]: `routingTable` can be null on publisher (on debugging there is no table corresponding to `shardId.getIndexName(`)) not sure if it's related to test setup or normal situation. 

On master when polling is done against such state, we fail on   
`var indexShard = indicesService.indexServiceSafe(shardId.getIndex()).getShard(shardId.id()); `with `IndexNotFoundException` - we don't fail listener but it's get caught somewhere in `InboundHandler` and then listener fails there. 

this btw means we can just throw on async execution as well - but I decided to shortcut a bit instead of relying on further handling in the `InboundHandler`
